File: algorithms/selection_sort/selection_sort.c
#include<stdio.h>
#include<stdlib.h>


void selection_sort(int *arr, int size) {
    int i = 0;

    while (i < size - 1) {
        int key = arr[i];
        for (int j = i + 1; j < size; j++) {
            if (arr[j] < key) {
                int tmp = key;
                key = arr[j];
                arr[j] = tmp;
            }
        }
        arr[i] = key;
        i++;
    }
}


void selection_sort_v2(int *arr, int size) {
    int i = 0;
    for (int i = 0; i < size - 1; i++) {
        int key = i;
        for (int j = i + 1; j < size; j++) {
            if (arr[j] < arr[key])
                key = j;
        }

        int tmp = arr[i];
        arr[i] = arr[key];
        arr[key] = tmp;
    }
}


int main(int argc, char *argv[]) {
    int n = argc - 1;
    int *arr = malloc(n * sizeof(int));

    for (int i = 0; i < n; i++) {
        char *endPtr;
        arr[i] = strtol(argv[i + 1], &endPtr, 10);
    }

    printf("Original array:\n");
    for (int i = 0; i < n; i++)
        printf("%d\t", arr[i]);

    selection_sort(arr, n);
    // selection_sort_v2(arr, n);

    printf("The new array is:\n");
    for (int i = 0; i < n; i++) {
        printf("%d\t", arr[i]);
    }

    free(arr);
    return 0;
}

File: algorithms/selection_sort/selection_sort.py
from random import randint


def selection_sort(arr: list[int]):
    while i < len(arr) - 1:
        key = arr[i]
        for j in range(i + 1, len(arr)):
            if arr[j] > key:
                tmp = key
                key = arr[j]
                arr[j] = tmp

        arr[j] = key
        i += 1


def selection_sort_v2(arr: list[int]):
    for i in range(len(arr) - 1):
        key = i
        for j in range(i + 1, len(arr)):
            if arr[j] < key:
                key = j
        tmp = arr[i]
        arr[i] = arr[key]
        arr[key] = tmp


def main():
    arr = [randint(0, 100) for _ in range(10)]
    print("The original array is:")
    print(arr)

    selection_sort(arr)
    # uncomment the following line for another version
    # selection_sort_v2(arr)

    print("The new array is:")
    print(arr)

if __name__ == "__main__":
    main()
